### PR TITLE
Added struct definition detection for ecto schemas

### LINKED ISSUE #695: Index `Ecto` schema struct
For example, when we define a module like:

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :name, :string
    field :age, :integer, default: 0
    field :password, :string, redact: true
    has_many :posts, Post
  end
end
```

and we can't go to the definition when the cursor at `%Use|r{}` right now, it is pretty annoying.

* https://hexdocs.pm/ecto/Ecto.Schema.html#schema/2
* https://hexdocs.pm/ecto/Ecto.Schema.html#embedded_schema/1
* https://hexdocs.pm/ecto/Ecto.Schema.html#embeds_many/3-inline-embedded-schema
* https://hexdocs.pm/ecto/Ecto.Schema.html#embeds_one/3

### PR DESCRIPTION
**Pull request by scohen**
We now have an extractor that emits a struct definition where ecto structs are defined.

Fixes #695

### CODE REVIEW

File: apps/remote_control/lib/lexical/remote_control/search/indexer/extractors/ecto_schema.ex [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+  defp block_form?(ast) do
+    {_, result} =
+      Macro.prewalk(ast, false, fn
+        {:__block__, _, [:do]}, false ->
+          {nil, true}
+
+        ast, acc ->
+          {ast, acc}
+      end)
```
Comment (zachallaun) [inline apps/remote_control/lib/lexical/remote_control/search/indexer/extractors/ecto_schema.ex:93]: Something that I think should go in a different PR:

We should add an `Ast.prewalk_while/3` that's similar to Sourceror's [`Zipper.traverse_while/3`](https://hexdocs.pm/sourceror/Sourceror.Zipper.html#traverse_while/3) in that the function can return `{:cont, ast, acc}`, `{:skip, ast, acc}`, or `{:halt, ast, acc}` so that we can short-circuit instead of walking the whole tree in cases like this.

```elixir
{_, result} =
  Ast.prewalk_while(ast, false, fn
    {:__block__, _, [:do]}, _ ->
      {:halt, nil, true}

    ast, acc ->
      {:cont, ast, acc}
  end)
```

File: apps/remote_control/test/lexical/remote_control/search/indexer/extractors/ecto_schema_test.exs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+      {:ok, [struct], _doc} =
+        ~q[
+         defmodule MySchema do
+           alias Ecto.Schema , as: SCM
+           use SCM
+           schema "my_schema" do
+             field :last_name, :string
+           end
+        end
+        ]
```
Comment (zachallaun) [inline apps/remote_control/test/lexical/remote_control/search/indexer/extractors/ecto_schema_test.exs:33]: There's a lot of inconsistent indentation in the code examples in these tests. Can we clean it up?

```suggestion
        ~q[
          defmodule MySchema do
            alias Ecto.Schema , as: SCM
            use SCM
            schema "my_schema" do
              field :last_name, :string
            end
          end
        ]
```

**Review by zachallaun — CHANGES_REQUESTED** [review]

File: apps/remote_control/test/lexical/remote_control/search/indexer/extractors/ecto_schema_test.exs (hunk 2 above)
Comment (scohen) [reply apps/remote_control/test/lexical/remote_control/search/indexer/extractors/ecto_schema_test.exs:33]: most of the inconsistent indentation in the expected cases is due to how extract works. Since it's pulling out a range that has a surrounding indentation, but the range pulls out the context so it starts with no indentation, however, the next line still contains the context's indentation.

File: apps/remote_control/lib/lexical/remote_control/search/indexer/extractors/ecto_schema.ex (hunk 1 above)
Comment (scohen) [reply apps/remote_control/lib/lexical/remote_control/search/indexer/extractors/ecto_schema.ex:93]: how do you stop a prewalk though?

Comment (zachallaun) [reply apps/remote_control/lib/lexical/remote_control/search/indexer/extractors/ecto_schema.ex:93]: I'm not sure I understand the question.

Comment (scohen) [reply apps/remote_control/lib/lexical/remote_control/search/indexer/extractors/ecto_schema.ex:93]: Pre walk is implemented via Macro.prewalk, I don't know how to stop it, and it will iterate through all the ast nodes eventually

File: apps/remote_control/test/lexical/remote_control/search/indexer/extractors/ecto_schema_test.exs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+      assert [] == results
+    end
+
+    test "consisting of a single field" do
+      {:ok, [struct], doc} =
+        ~q[
+         defmodule MySchema do
+           use Ecto.Schema
+           embedded_schema "my_schema" do
```
Comment (scottming) [inline apps/remote_control/test/lexical/remote_control/search/indexer/extractors/ecto_schema_test.exs:119]:
```suggestion
           embedded_schema  do
```
